File: Populating-Next-Right-Pointers-in-Each-Node-II/Populating-Next-Right-Pointers-in-Each-Node-II.py


# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree node
    # @return nothing
    def connect(self, root):
        if root==None:
            return
        if root.left!=None:
            root.left.next = root.right
            if root.right==None:
                node = root.next
                while node!=None and root.left.next==None:
                    root.left.next = node.left
                    if node.left==None:
                        root.left.next = node.right
                    node = node.next
        if root.right!=None:
            node = root.next
            while node!=None and root.right.next==None:
                root.right.next = node.left
                if node.left==None:
                    root.right.next = node.right
                node = node.next
        self.connect(root.right)
        self.connect(root.left)
 

File: Populating-Next-Right-Pointers-in-Each-Node-II/Populating-Next-Right-Pointers-in-Each-Node-II.cpp

/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        while(root)
        {
            TreeLinkNode *phead=NULL;
            TreeLinkNode *prev=NULL;
            while(root)
            {
                if(root->left)
                {
                    if(prev)
                        prev->next=root->left;
                    else
                        phead=root->left;
                    prev=root->left;
                }
                if(root->right)
                {
                    if(prev)
                        prev->next=root->right;
                    else
                        phead=root->right;
                    prev=root->right;
                }
                root=root->next;
            }
            root=phead;
        }
    }
};


 